applications/asset_tracker/src/main.c

```diff
@@ -1447,22 +1654,39 @@ void main(void)
 		}
 
 		if ((fds[0].revents & POLLNVAL) == POLLNVAL) {
-			if (atomic_get(&reconnect_to_cloud)) {
-				k_delayed_work_cancel(&cloud_reboot_work);
-				LOG_INF("Attempting reconnect...");
+			if ((atomic_get(&cloud_association) ==
+					CLOUD_ASSOCIATION_STATE_RECONNECT) ||
+				(atomic_get(&carrier_requested_disconnect))) {
+				reconnecting = true;
+				reconnect_delay_s = 10;
 				goto connect;
 			}
 			LOG_ERR("Socket error: POLLNVAL");
 			LOG_ERR("The cloud socket was unexpectedly closed.");
 			error_handler(ERROR_CLOUD, -EIO);
-			return;
 		}
 
 		if ((fds[0].revents & POLLHUP) == POLLHUP) {
+			if ((atomic_get(&cloud_connect_attempts) == 1) &&
```

Comment:
It's not the purpose of this PR to fix every situation, but hopefully this can be changed later to reconnect from a POLLHUP due to NAT timeout as well.

Reply:
agreed.  once we have a method to reveal the cause of a POLLHUP, that can be done.